### PR TITLE
Automatic carousel rotation

### PR DESCRIPTION
**Pull request by ouchadam**
Part of #4584

- Adds automatic 5 second rotation to the Ftue auth carousel
- Also included a super handy extension from [SO](https://stackoverflow.com/questions/57505875/change-viewpager2-scroll-speed-when-sliding-programmatically/59235979#59235979) to control the view pager page transition speed as it's too fast out of the box
- Fixes the carousel direction for RTL locales

| GIF | RTL
| --- | --- |
|device-2021-12-16-144427|after-rtl

### CODE REVIEW

**Review by bmarty — APPROVED** [review]
LGTM.
Is it working fine for RTL languages (I did not test it)?

File: vector/src/main/java/im/vector/app/core/extensions/Integer.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.core.extensions
+
+fun Int.incrementAndWrap(max: Int, min: Int = 0, amount: Int = 1) = if (this == max) min else this + amount
```
[the remark this answers is not in the document]
Comment (ouchadam) [reply vector/src/main/java/im/vector/app/core/extensions/Integer.kt]: https://github.com/vector-im/element-android/pull/4740/commits/8663714de4e69da218fbc969f25129962b2ae90a made the helper specific to increments of 1 to keep things simpler

File: vector/src/main/java/im/vector/app/features/onboarding/ftueauth/FtueAuthSplashCarouselFragment.kt [hunk 2]
```diff
@@ -110,7 +110,7 @@ class FtueAuthSplashCarouselFragment @Inject constructor(
         views.splashCarousel.apply {
             val itemCount = carouselAdapter.itemCount
             tickerFlow(lifecycleScope, delayMillis = CAROUSEL_ROTATION_DELAY_MS)
-                    .onEach { setCurrentItem(currentItem.incrementAndWrap(max = itemCount - 1), duration = CAROUSEL_TRANSITION_TIME_MS) }
+                    .onEach { setCurrentItem(currentItem.incrementByOneAndWrap(max = itemCount - 1), duration = CAROUSEL_TRANSITION_TIME_MS) }
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/onboarding/ftueauth/FtueAuthSplashCarouselFragment.kt:113]: I also like the simple form `setCurrentItem((currentItem + 1) % itemCount)`

**Comment by bmarty** [comment]
Just a small remark while testing it:
Maybe disable the automatic pagination once the user had manually change the page.

File: library/ui-styles/src/main/res/drawable/bg_carousel_page_1.xml [hunk 3]
```diff
@@ -1,6 +1,7 @@
 <?xml version="1.0" encoding="utf-8"?>
 <shape xmlns:android="http://schemas.android.com/apk/res/android">
     <gradient
+        android:angle="@integer/rtl_mirror_flip"
```
Comment (ouchadam) [inline library/ui-styles/src/main/res/drawable/bg_carousel_page_1.xml:4]: @bmarty the rtl gradient is fixed by making use of the `rtl_mirror_flip`, ltr = 0 rtl = 180

File: vector/src/main/java/im/vector/app/core/extensions/ViewPager2.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    val pxToDrag: Int = pagePxWidth * (item - currentItem)
+    val animator = ValueAnimator.ofInt(0, pxToDrag)
+    var previousValue = 0
+    val isRtl = this.layoutDirection == View.LAYOUT_DIRECTION_RTL
+
+    animator.addUpdateListener { valueAnimator ->
+        val currentValue = valueAnimator.animatedValue as Int
+        val currentPxToDrag = (currentValue - previousValue).toFloat()
+        kotlin.runCatching {
+            when {
```
Comment (ouchadam) [inline vector/src/main/java/im/vector/app/core/extensions/ViewPager2.kt:41]: @bmarty also updated the fake scrolling to take into account if we need to fake scroll ltr or rtl

**Comment by ouchadam** [comment]
> Just a small remark while testing it: Maybe disable the automatic pagination once the user had manually change the page.

checking in with product to see what they think, I agree it could be quite annoying to manually select a page only for it to keep changing

**Comment by bmarty** [comment]
Ok, thanks. On my side I really want to stop animation on user interaction. Especially the timer is not reset so the user can go back to the previous page to finish reading it and the page change again automatically just after, this is not nice.

**Comment by bmarty** [comment]
Another note to discuss with product/design: some of the images contains wording, which is not ideal for localisation. Can we avoid that? FTR it is also possible to add images per languages, but I would prefer to avoid doing that if possible

image

image

CC @daniellekirkwood

**Comment by ouchadam** [comment]
@bmarty we've already done it! :sunglasses:  https://www.figma.com/file/xdV4PuI3DlzA1EiBvbrggz/Login-Flow-v2?node-id=2235%3A41935 

just need to update the temporary images with the updated versions

**Comment by bmarty** [comment]
Update for RTL tested OK 💯 (build 5364).

But I had a crash when fast clicking on the TabLayout `carouselIndicator`:

```bash
    Thread: main, Exception: java.lang.IllegalStateException: Cannot change current item when ViewPager2 is fake dragging
        at androidx.viewpager2.widget.ViewPager2.setCurrentItem(ViewPager2.java:604)
        at com.google.android.material.tabs.TabLayoutMediator$ViewPagerOnTabSelectedListener.onTabSelected(TabLayoutMediator.java:268)
```

**Comment by ouchadam** [comment]
fix for the crash whilst our fake dragging is taking place https://github.com/vector-im/element-android/pull/4740/commits/a0bda0282498daeb2ee731dba36a6a982accedc4

updated the automatic transition behaviour to only schedule once the page settings, rather than on a constant ticker, matches the iOS behaviour https://github.com/vector-im/element-android/pull/4740/commits/70c82443ee235811b4e29a08f7fe20ddf2f42f15 (whilst we wait for confirmation about whether the automatic transitions should be fully disabled on user interaction)

**Review by bmarty — APPROVED** [review]
Thanks for the update

**Pull request merged**